cmd/snap-seccomp/main.go

```diff
@@ -0,0 +1,522 @@
+	}
+
+	return strconv.ParseUint(token, 10, 64)
+}
+
+func parseLine(line string, secFilter *seccomp.ScmpFilter) error {
+	line = strings.TrimSpace(line)
+
+	// ignore comments and empty lines
+	if strings.HasPrefix(line, "#") || line == "" {
+		return nil
+	}
+
+	// regular line
+	tokens := strings.Fields(line)
+
+	// fish out syscall
+	secSyscall, err := seccomp.GetSyscallFromName(tokens[0])
+	if err != nil {
+		// FIXME: add structed error to libseccomp-golang
+		if err.Error() == "could not resolve name to syscall" {
+			return nil
+		}
+		return fmt.Errorf("cannot resolve name: %s", err)
+	}
+
+	var conds []seccomp.ScmpCondition
+	if err != nil {
+		return fmt.Errorf("cannot create new filter: %s", err)
+	}
+
+	for pos, arg := range tokens[1:] {
+		var cmpOp seccomp.ScmpCompareOp
+		var value uint64
+		var err error
+
+		if pos >= ScArgsMaxlength {
+			return fmt.Errorf("too many tokens (%d) in line %q", pos, line)
+		}
+
+		if arg == "-" {
+			continue
+		}
+
+		if strings.HasPrefix(arg, ">=") {
+			cmpOp = seccomp.CompareGreaterEqual
+			value, err = readNumber(arg[2:])
+		} else if strings.HasPrefix(arg, "<=") {
+			cmpOp = seccomp.CompareLessOrEqual
+			value, err = readNumber(arg[2:])
+		} else if strings.HasPrefix(arg, "!") {
+			cmpOp = seccomp.CompareNotEqual
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, "<") {
+			cmpOp = seccomp.CompareLess
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, ">") {
+			cmpOp = seccomp.CompareGreater
+			value, err = readNumber(arg[1:])
+		} else if strings.HasPrefix(arg, "|") {
+			cmpOp = seccomp.CompareMaskedEqual
+			value, err = readNumber(arg[1:])
+		} else {
+			cmpOp = seccomp.CompareEqual
+			value, err = readNumber(arg)
+		}
+		if err != nil {
+			return fmt.Errorf("cannot parse token %q (line %q)", arg, line)
+		}
+
+		var scmpCond seccomp.ScmpCondition
+		if cmpOp == seccomp.CompareMaskedEqual {
+			scmpCond, err = seccomp.MakeCondition(uint(pos), cmpOp, value, value)
+		} else {
+			scmpCond, err = seccomp.MakeCondition(uint(pos), cmpOp, value)
+		}
+		if err != nil {
+			return fmt.Errorf("cannot parse line %q: %s", line, err)
+		}
+		conds = append(conds, scmpCond)
+	}
+
+	// FIXME: why do we need this fallback?
+	if err = secFilter.AddRuleConditionalExact(secSyscall, seccomp.ActAllow, conds); err != nil {
+		err = secFilter.AddRuleConditional(secSyscall, seccomp.ActAllow, conds)
+	}
+
+	return err
+}
+
+// For architectures that support a compat architecture, when the
+// kernel and userspace match, add the compat arch, otherwise add
+// the kernel arch to support the kernel's arch (eg, 64bit kernels with
+// 32bit userspace).
+func addSecondaryArches(secFilter *seccomp.ScmpFilter) error {
+	switch arch.UbuntuArchitecture() {
+	case "amd64":
+		return secFilter.AddArch(seccomp.ArchX86)
+	case "arm64":
+		return secFilter.AddArch(seccomp.ArchARM)
+	case "ppc64el":
+		return secFilter.AddArch(seccomp.ArchPPC)
+	}
+	return nil
```

Comment:
To give a practical example, most of the time snaps specify the architecture for the system they are destined for and the normal case is kernel, userspace and snap all match (ie, i386 or amd64) and there are no compat syscalls. Indeed, this was all snappy supported at first.

Some 64 bit systems support running 32 bit code on 64 bit systems, eg, a `wine` snap might ship in its amd64 snap *both* 64 bit and 32 bit binaries, and the 32 bit binaries will use compat syscalls. This is case '2'.

A brand may decide to create a model with a 64 bit kernel and a 32 bit userspace. While most snaps would simply use the 32bit architecture (eg, i386), a specialized (perhaps brand) snap on this system might contain some code that is 64 bit to take advantage of the kernel capabilities and this code would need to use the syscalls for the kernel architecture. This is case '3'.